api-docs/generator.py

```diff
@@ -1,40 +1,139 @@
+APIDOC_CONTAINERS = ["postgres", "redis", "clickhouse", "snuba", "relay", "reverse_proxy"]
+devservices_settings = {
+    container_name: SENTRY_DEVSERVICES[container_name] for container_name in APIDOC_CONTAINERS
+}
+
+apidoc_containers_overrides = {
+    "postgres": {"environment": {"POSTGRES_DB": "sentry_api_docs"}, "volumes": None},
+    "redis": {"volumes": None},
+    "clickhouse": {"ports": None, "volumes": None, "only_if": None},
+    "snuba": {
+        "pull": None,
+        "command": ["devserver", "--no-workers"],
+        "environment": {
+            "CLICKHOUSE_HOST": namespace + "_clickhouse",
+            "DEFAULT_BROKERS": None,
+            "REDIS_HOST": namespace + "_redis",
+        },
+        "volumes": None,
+        "only_if": None,
+    },
+    "relay": {"pull": None, "volumes": None, "only_if": None, "with_devserver": None},
+    "reverse_proxy": {"volumes": None, "only_if": None, "with_devserver": None},
+}
+
+
+def deep_merge(defaults, overrides):
+    """
+    Deep merges two dictionaries.
+
+    If the value is None in `overrides`, that key-value pair will not show up in the final result.
+    """
+    merged = {}
+    for key in defaults:
+        if isinstance(defaults[key], dict):
+            if key not in overrides:
+                merged[key] = defaults[key]
+            elif overrides[key] is None:
+                continue
+            elif isinstance(overrides[key], dict):
+                merged[key] = deep_merge(defaults[key], overrides[key])
+            else:
+                raise Exception("Types must match")
+        elif key in overrides and overrides[key] is None:
+            continue
+        elif key in overrides:
+            merged[key] = overrides[key]
+        else:
+            merged[key] = defaults[key]
+    return merged
+
+
+containers = deep_merge(devservices_settings, apidoc_containers_overrides)
+
+
+# Massage our list into some shared settings instead of repeating
+# it for each definition.
+for name, options in containers.items():
+    options["network"] = namespace
+    options["detach"] = True
+    options["name"] = namespace + "_" + name
+    containers[name] = options
+
+# Pull all of our unique images once.
+pulled = set()
+
+for name, options in containers.items():
+    if options["image"] not in pulled:
+        click.secho("> Pulling image '%s'" % options["image"], err=True, fg="green")
+        client.images.pull(options["image"])
+        pulled.add(options["image"])
+
+
+# Run each of our containers, if found running already, delete first
+# and create new. We never want to reuse.
+for name, options in containers.items():
+    try:
+        container = client.containers.get(options["name"])
+    except docker.errors.NotFound:
+        pass
+    else:
+        container.stop()
+        container.remove()
+
+    click.secho("> Creating '%s' container" % options["name"], err=True, fg="yellow")
+    client.containers.run(**options)
```

Comment:
This will run containers even when you only want to get usage information with `--help`. You might want to defer this until inside of main().